File: src/opentrace_ml/map_matching.py
"""Model-independent contracts for matching private traces to a road network."""

from __future__ import annotations

import math
import re
from collections.abc import Callable, Iterable
from dataclasses import dataclass
from typing import Protocol, runtime_checkable

from .models import GeoPoint
from .trace import PreparedTrace

_TRIP_PSEUDONYM = re.compile(r"trip_[0-9a-f]{64}")


@dataclass(frozen=True, slots=True)
class RawMapMatch:
    """One adapter result before its source point is attached."""

    source_index: int
    matched_point: GeoPoint | None = None
    confidence: float = 0.0
    edge_id: str | None = None

    def __post_init__(self) -> None:
        if (
            not isinstance(self.source_index, int)
            or isinstance(self.source_index, bool)
            or self.source_index < 0
        ):
            raise ValueError("source_index must be a non-negative integer")
        if not math.isfinite(self.confidence) or not 0.0 <= self.confidence <= 1.0:
            raise ValueError("Map-match confidence must be finite and between 0 and 1")
        if self.matched_point is None:
            if self.confidence != 0.0 or self.edge_id is not None:
                raise ValueError("Unmatched observations cannot have confidence or an edge ID")
            return
        if self.confidence <= 0.0:
            raise ValueError("Matched observations must have positive confidence")
        if self.edge_id is None or not self.edge_id.strip():
            raise ValueError("Matched observations require a non-empty edge ID")


@dataclass(frozen=True, slots=True)
class MapMatchObservation:
    """A source trace point and its optional road-network match."""

    source_index: int
    source_point: GeoPoint
    matched_point: GeoPoint | None
    confidence: float
    edge_id: str | None

    def __post_init__(self) -> None:
        RawMapMatch(
            source_index=self.source_index,
            matched_point=self.matched_point,
            confidence=self.confidence,
            edge_id=self.edge_id,
        )
        if (
            self.matched_point is not None
            and self.matched_point.timestamp_seconds != self.source_point.timestamp_seconds
        ):
            raise ValueError("Matched points must preserve the source timestamp")

    @property
    def is_matched(self) -> bool:
        """Whether the source point was matched to a road-network edge."""

        return self.matched_point is not None


@dataclass(frozen=True, slots=True)
class MapMatchSegment:
    """A contiguous matched edge or unmatched run without raw coordinates."""

    start_source_index: int
    end_source_index: int
    matched: bool
    edge_id: str | None
    mean_confidence: float

    def __post_init__(self) -> None:
        if self.start_source_index < 0 or self.end_source_index < self.start_source_index:
            raise ValueError("Map-match segment index range is invalid")
        if not math.isfinite(self.mean_confidence) or not 0.0 <= self.mean_confidence <= 1.0:
            raise ValueError("Segment confidence must be finite and between 0 and 1")
        if self.matched:
            if self.edge_id is None or not self.edge_id.strip():
                raise ValueError("Matched segments require a non-empty edge ID")
            if self.mean_confidence <= 0.0:
                raise ValueError("Matched segments must have positive confidence")
        elif self.edge_id is not None or self.mean_confidence != 0.0:
            raise ValueError("Unmatched segments cannot have confidence or an edge ID")


@dataclass(frozen=True, slots=True)
class MapMatchResult:
    """Ordered map-match output for one pseudonymized private trace."""

    trip_id: str
    observations: tuple[MapMatchObservation, ...]

    def __post_init__(self) -> None:
        if _TRIP_PSEUDONYM.fullmatch(self.trip_id) is None:
            raise ValueError("Map-match results require an OpenTrace trip pseudonym")
        if not self.observations:
            raise ValueError("Map-match results cannot be empty")
        indices = tuple(observation.source_index for observation in self.observations)
        if indices != tuple(range(len(self.observations))):
            raise ValueError("Map-match observations must cover every source point in order")

    @property
    def matched_count(self) -> int:
        return sum(observation.is_matched for observation in self.observations)

    @property
    def unmatched_count(self) -> int:
        return len(self.observations) - self.matched_count

    @property
    def unmatched_ratio(self) -> float:
        return self.unmatched_count / len(self.observations)

    @property
    def segments(self) -> tuple[MapMatchSegment, ...]:
        """Group consecutive observations by matched edge or unmatched status."""

        segments: list[MapMatchSegment] = []
        start = 0
        while start < len(self.observations):
            first = self.observations[start]
            key = (first.is_matched, first.edge_id)
            end = start + 1
            while end < len(self.observations):
                current = self.observations[end]
                if (current.is_matched, current.edge_id) != key:
                    break
                end += 1
            run = self.observations[start:end]
            segments.append(
                MapMatchSegment(
                    start_source_index=start,
                    end_source_index=end - 1,
                    matched=first.is_matched,
                    edge_id=first.edge_id,
                    mean_confidence=sum(item.confidence for item in run) / len(run),
                )
            )
            start = end
        return tuple(segments)


@runtime_checkable
class MapMatcher(Protocol):
    """Minimal interface implemented by road-network matcher adapters."""

    def match(self, trace: PreparedTrace) -> MapMatchResult:
        """Match a consented, pseudonymized trace without publishing its points."""


@dataclass(slots=True)
class CallableMapMatcher:
    """Wrap a deterministic callable in the OpenTrace map-matcher contract."""

    matcher: Callable[[PreparedTrace], Iterable[RawMapMatch]]

    def match(self, trace: PreparedTrace) -> MapMatchResult:
        if not isinstance(trace, PreparedTrace):
            raise TypeError("Map matchers require a PreparedTrace")

        raw_matches = tuple(self.matcher(trace))
        if len(raw_matches) != len(trace.points):
            raise ValueError("A matcher must return one observation for every source point")

        observations: list[MapMatchObservation] = []
        for expected_index, item in enumerate(raw_matches):
            if not isinstance(item, RawMapMatch):
                raise TypeError("Map matcher adapters must return RawMapMatch objects")
            if item.source_index != expected_index:
                raise ValueError("Map matcher output must preserve source-point order")
            observations.append(
                MapMatchObservation(
                    source_index=item.source_index,
                    source_point=trace.points[item.source_index],
                    matched_point=item.matched_point,
                    confidence=item.confidence,
                    edge_id=item.edge_id,
                )
            )

        return MapMatchResult(trip_id=trace.trip_id, observations=tuple(observations))

File: go/mapmatching.go
package opentrace

import (
	"context"
	"errors"
	"fmt"
	"strings"
)

// RawMapMatch is one adapter response before its private source point is
// attached. A nil MatchedPoint represents an unmatched observation.
type RawMapMatch struct {
	SourceIndex  int       `json:"source_index"`
	MatchedPoint *GeoPoint `json:"matched_point,omitempty"`
	Confidence   float64   `json:"confidence"`
	EdgeID       string    `json:"edge_id,omitempty"`
}

// Validate checks the adapter-facing map-match contract.
func (match RawMapMatch) Validate() error {
	if match.SourceIndex < 0 {
		return errors.New("opentrace: source index cannot be negative")
	}
	if !isFinite(match.Confidence) || match.Confidence < 0 || match.Confidence > 1 {
		return errors.New("opentrace: map-match confidence must be finite and between 0 and 1")
	}
	if match.MatchedPoint == nil {
		if match.Confidence != 0 || match.EdgeID != "" {
			return errors.New("opentrace: unmatched observations cannot have confidence or an edge ID")
		}
		return nil
	}
	if err := match.MatchedPoint.Validate(); err != nil {
		return fmt.Errorf("opentrace: invalid matched point: %w", err)
	}
	if match.Confidence <= 0 {
		return errors.New("opentrace: matched observations must have positive confidence")
	}
	if strings.TrimSpace(match.EdgeID) == "" {
		return errors.New("opentrace: matched observations require an edge ID")
	}
	return nil
}

// MapMatchObservation joins a private source point with an optional match.
type MapMatchObservation struct {
	SourceIndex  int       `json:"source_index"`
	SourcePoint  GeoPoint  `json:"source_point"`
	MatchedPoint *GeoPoint `json:"matched_point,omitempty"`
	Confidence   float64   `json:"confidence"`
	EdgeID       string    `json:"edge_id,omitempty"`
}

// IsMatched reports whether the observation has a road-network match.
func (observation MapMatchObservation) IsMatched() bool {
	return observation.MatchedPoint != nil
}

func (observation MapMatchObservation) validate() error {
	raw := RawMapMatch{
		SourceIndex:  observation.SourceIndex,
		MatchedPoint: observation.MatchedPoint,
		Confidence:   observation.Confidence,
		EdgeID:       observation.EdgeID,
	}
	if err := raw.Validate(); err != nil {
		return err
	}
	if err := observation.SourcePoint.Validate(); err != nil {
		return fmt.Errorf("opentrace: invalid source point: %w", err)
	}
	if observation.MatchedPoint != nil &&
		observation.MatchedPoint.TimestampSeconds != observation.SourcePoint.TimestampSeconds {
		return errors.New("opentrace: matched points must preserve the source timestamp")
	}
	return nil
}

// MapMatchSegment summarizes a contiguous matched edge or unmatched run
// without exposing coordinates.
type MapMatchSegment struct {
	StartSourceIndex int     `json:"start_source_index"`
	EndSourceIndex   int     `json:"end_source_index"`
	Matched          bool    `json:"matched"`
	EdgeID           string  `json:"edge_id,omitempty"`
	MeanConfidence   float64 `json:"mean_confidence"`
}

// MapMatchResult is ordered output for one pseudonymized trace.
type MapMatchResult struct {
	TripID       string                `json:"trip_id"`
	Observations []MapMatchObservation `json:"observations"`
}

// Validate checks pseudonym, ordering, and one-result-per-point invariants.
func (result MapMatchResult) Validate() error {
	if !validTripPseudonym(result.TripID) {
		return ErrInvalidPseudonym
	}
	if len(result.Observations) == 0 {
		return errors.New("opentrace: map-match result cannot be empty")
	}
	for i, observation := range result.Observations {
		if observation.SourceIndex != i {
			return errors.New("opentrace: map-match observations must preserve source-point order")
		}
		if err := observation.validate(); err != nil {
			return fmt.Errorf("opentrace: observation %d: %w", i, err)
		}
	}
	return nil
}

// MatchedCount returns the number of matched observations in O(n).
func (result MapMatchResult) MatchedCount() int {
	count := 0
	for _, observation := range result.Observations {
		if observation.IsMatched() {
			count++
		}
	}
	return count
}

// UnmatchedCount returns the number of unmatched observations in O(n).
func (result MapMatchResult) UnmatchedCount() int {
	return len(result.Observations) - result.MatchedCount()
}

// UnmatchedRatio returns the unmatched fraction in O(n).
func (result MapMatchResult) UnmatchedRatio() float64 {
	if len(result.Observations) == 0 {
		return 0
	}
	return float64(result.UnmatchedCount()) / float64(len(result.Observations))
}

// Segments groups consecutive observations by matched edge or unmatched
// status in O(n) time.
func (result MapMatchResult) Segments() []MapMatchSegment {
	segments := make([]MapMatchSegment, 0)
	for start := 0; start < len(result.Observations); {
		first := result.Observations[start]
		end := start + 1
		sum := first.Confidence
		for end < len(result.Observations) {
			current := result.Observations[end]
			if current.IsMatched() != first.IsMatched() || current.EdgeID != first.EdgeID {
				break
			}
			sum += current.Confidence
			end++
		}
		segments = append(segments, MapMatchSegment{
			StartSourceIndex: start,
			EndSourceIndex:   end - 1,
			Matched:          first.IsMatched(),
			EdgeID:           first.EdgeID,
			MeanConfidence:   sum / float64(end-start),
		})
		start = end
	}
	return segments
}

// MapMatcher is the coarse adapter boundary for Valhalla, OSRM, or another
// road-network matcher. Implementations should honor context cancellation.
type MapMatcher interface {
	Match(context.Context, PreparedTrace) (MapMatchResult, error)
}

// MapMatcherFunc adapts a function to MapMatcher.
type MapMatcherFunc func(context.Context, PreparedTrace) (MapMatchResult, error)

// Match calls the wrapped function.
func (function MapMatcherFunc) Match(ctx context.Context, trace PreparedTrace) (MapMatchResult, error) {
	if function == nil {
		return MapMatchResult{}, errors.New("opentrace: map matcher function is nil")
	}
	return function(ctx, trace)
}

// NewMapMatchResult attaches a match to every point of a PreparedTrace. It
// copies matched points so callers cannot mutate adapter-owned values later.
func NewMapMatchResult(trace PreparedTrace, matches []RawMapMatch) (MapMatchResult, error) {
	points := trace.Points()
	if len(matches) != len(points) {
		return MapMatchResult{}, errors.New("opentrace: matcher must return one observation per source point")
	}
	observations := make([]MapMatchObservation, len(matches))
	for i, match := range matches {
		if err := match.Validate(); err != nil {
			return MapMatchResult{}, fmt.Errorf("opentrace: raw match %d: %w", i, err)
		}
		if match.SourceIndex != i {
			return MapMatchResult{}, errors.New("opentrace: map matcher output must preserve source-point order")
		}
		var matchedPoint *GeoPoint
		if match.MatchedPoint != nil {
			copyOfPoint := *match.MatchedPoint
			matchedPoint = &copyOfPoint
		}
		observations[i] = MapMatchObservation{
			SourceIndex:  i,
			SourcePoint:  points[i],
			MatchedPoint: matchedPoint,
			Confidence:   match.Confidence,
			EdgeID:       match.EdgeID,
		}
	}
	result := MapMatchResult{TripID: trace.TripID(), Observations: observations}
	if err := result.Validate(); err != nil {
		return MapMatchResult{}, err
	}
	return result, nil
}

// RunMapMatcher invokes an adapter and validates untrusted output.
func RunMapMatcher(ctx context.Context, matcher MapMatcher, trace PreparedTrace) (MapMatchResult, error) {
	if ctx == nil {
		return MapMatchResult{}, errors.New("opentrace: context is nil")
	}
	if err := ctx.Err(); err != nil {
		return MapMatchResult{}, err
	}
	if !validTripPseudonym(trace.TripID()) || trace.Len() < 2 {
		return MapMatchResult{}, errors.New("opentrace: map matchers require a prepared trace")
	}
	if matcher == nil {
		return MapMatchResult{}, errors.New("opentrace: map matcher is nil")
	}
	result, err := matcher.Match(ctx, trace)
	if err != nil {
		return MapMatchResult{}, err
	}
	if err := ctx.Err(); err != nil {
		return MapMatchResult{}, err
	}
	if err := result.Validate(); err != nil {
		return MapMatchResult{}, fmt.Errorf("opentrace: invalid map matcher output: %w", err)
	}
	if result.TripID != trace.TripID() || len(result.Observations) != trace.Len() {
		return MapMatchResult{}, errors.New("opentrace: map matcher output does not belong to the input trace")
	}
	for i, observation := range result.Observations {
		if observation.SourcePoint != trace.points[i] {
			return MapMatchResult{}, errors.New("opentrace: map matcher changed a source point")
		}
	}
	return result, nil
}
